### PR TITLE
KAFKA-2105: add topic null check to KafkaProducer.partitionsFor method

### PR DESCRIPTION
**Pull request by omkreddy**

### CODE REVIEW

**Comment by omkreddy** [comment]
not sure this will create any compatibility issue. 
cc @ijuma

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6891/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6877/
Test PASSed (JDK 8 and Scala 2.12).

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 1]
```diff
@@ -914,7 +915,7 @@ public void flush() {
     @Override
     public List<PartitionInfo> partitionsFor(String topic) {
         try {
-            return waitOnMetadata(topic, null, maxBlockTimeMs).cluster.partitionsForTopic(topic);
+            return waitOnMetadata(Utils.notNull(topic), null, maxBlockTimeMs).cluster.partitionsForTopic(topic);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: `Metadata.add` should also not allow a null `topic` either. Also, we provide an error message if a user provides a null `topic` here.

Comment (omkreddy) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: updated the PR with changes. pls check

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 2]
```diff
@@ -965,6 +965,10 @@ public void flush() {
     @Override
     public List<PartitionInfo> partitionsFor(String topic) {
         try {
+            if (topic == null) {
+                log.error("topic cannot be null");
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Hmm, there's no reason to log this. Also, the check should be outside the `try`. Something like the following would do:

```java
Objects.requireNonNull(topic, "topic cannot be null")
```

I'd use a similar pattern in `Metadata.add`.

Comment (omkreddy) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Thanks for the suggestion. much cleaner now. updated the PR.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 3]
```diff
@@ -464,4 +464,12 @@ public void testInterceptorPartitionSetOnTooLargeRecord() throws Exception {
 
         EasyMock.verify(interceptors);
     }
+
+    @Test(expected = NullPointerException.class)
+    public void testPartitionsForWithNullTopic() {
+        Properties props = new Properties();
+        props.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9000");
+        KafkaProducer<byte[], byte[]> producer = new KafkaProducer<>(props, new ByteArraySerializer(), new ByteArraySerializer());
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: We should close the producer.

Comment (omkreddy) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Thanks. Updated the PR

**Review by ijuma — APPROVED** [review]
LGTM, thanks. Merging to trunk and 0.11.0.

**Pull request closed without merging**